wzdx/tools/geospatial_tools.py

```diff
@@ -0,0 +1,126 @@
+import math
+
+import pyproj
+
+ROAD_DIRECTIONS_MAP = {0: "northbound", 1: "eastbound", 2: "southbound", 3: "westbound"}
+
+ROAD_ORIENTATIONS_MAP = {
+    "northbound": 0,
+    "eastbound": 90,
+    "southbound": 0,
+    "westbound": 90,
+}
+
+ROAD_ORIENTATIONS_DIRECTIONS_MAP = {
+    0: ["northbound", "southbound", "southbound", "northbound"],
+    90: ["eastbound", "westbound", "westbound", "eastbound"],
+}
+
+
+# function to get road direction by using geometry coordinates
+def get_road_direction_from_coordinates(coordinates):
+    if not coordinates or type(coordinates) != list or len(coordinates) < 2:
+        return "unknown"
+
+    try:
+        long_dif = coordinates[-1][0] - coordinates[0][0]
+        lat_dif = coordinates[-1][1] - coordinates[0][1]
+    except ValueError as e:
+        return "unknown"
+    except IndexError as e:
+        return "unknown"
+
+    if abs(long_dif) > abs(lat_dif):
+        if long_dif > 0:
+            direction = "eastbound"
+        else:
+            direction = "westbound"
+    elif lat_dif > 0:
+        direction = "northbound"
+    else:
+        direction = "southbound"
+
+    if lat_dif == 0 and long_dif == 0:
+        direction = "unknown"
+
+    return direction
+
+
+def get_heading_from_coordinates(coordinates):
+    """Return the heading between two long/lat coordinates"""
+    if not coordinates or type(coordinates) != list or len(coordinates) < 2:
+        return None
+
+    geodesic_pyproj = pyproj.Geod(ellps="WGS84")
+
+    fwd_heading, _, __ = geodesic_pyproj.inv(
+        coordinates[0][0], coordinates[0][1], coordinates[1][0], coordinates[1][1]
+    )
+
+    return fwd_heading % 360
+
+
+# unnecessarily condensed just because
+def get_direction_from_bearing(bearing):
+    return ROAD_DIRECTIONS_MAP[math.floor((bearing + 45) / 90) % 4]
+
+
+# unnecessarily condensed just because
+def get_closest_direction_from_bearing(bearing, road_orientation):
```

Comment:
What are we doing with this method? The ROAD_ORIENTATIONS_DIRECTIONS_MAP seems a bit complex.

Reply:
Added a docstring comment: Return the direction of the object, by snapping the bearing to the road orientation (allows reversals, e.g. northbound or southbound can be returned for a northbound road orientation)

I somewhat unnecessarily condensed this logic, but the unit tests more than sufficiently validate the functionality